File: baekjoon/단계 별로 풀어보기/스택/4949_균형잡힌_세상.py
import sys
input = sys.stdin.readline
IP = lambda: map(int, input().split())

def solve():
    while True:
        string = input().rstrip()
        if string == '.': break
        stack = []
        isit = True
        
        for s in string:
            if s == "(": stack.append(1)
            elif s == ")":
                if stack and stack[-1] == 1: stack.pop()
                else: isit = False; break
            elif s == "[": stack.append(2)
            elif s == "]":
                if stack and stack[-1] == 2: stack.pop()
                else: isit = False; break
        
        print("yes" if isit and not stack else "no")
    
if "__main__" == __name__: solve()

File: baekjoon/단계 별로 풀어보기/스택/4949_균형잡힌_세상.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <cmath>
#include <tuple>
#include <stack>
using namespace std;

int main()
{
	cin.tie(NULL);
	cout.tie(NULL);
	ios_base::sync_with_stdio(false);
	
	string S;
	while (true){
	    getline(cin, S);
	    if (S == ".") break;
	    stack<bool> st;
	    bool isit = true;
	    
	    for (auto &s : S){
	        if (s == '(') st.push(true);
	        else if (s == ')'){
	            if (!st.empty() && st.top()) st.pop();
	            else{
	                isit = false;
	                break;
	            }
	        }
	        else if (s == '[') st.push(false);
	        else if (s == ']'){
	            if (!st.empty() && !st.top()) st.pop();
	            else{
	                isit = false;
	                break;
	            }
	        }
	    }
	    cout << ((isit && st.empty()) ? "yes" : "no") << "\n";
	}
}